### DISCUSSION TITLE
OS/2 and OPENSTEP in emulation on Apple silicon

### DISCUSSION THREAD

**Discussion by mbert** [utmapp/UTM#3796] [UTM Help (macOS)]
I am a lover of exotic and legacy operating systems, and after I got my Apple silicon MBP one thing on my todo list was rescuing my favourite VMs after they could no longer be executed on VMWare Fusion.

I converted the disk images to qcow2 format, created VMs, took over the settings from the old VMs, imported the new disk images and went for it. This worked fine with VMs of more recent OSes, like Haiku OS. However I got stuck with OS/2 and OPENSTEP:

OPENSTEP (4.2) tells me "OPENSTEP boot1 v40.13.1" but gets stuck there. No error message, simply silence. 

OS/2 Warp Server 4 went into errors when booting until I selected an older hardware profile. After having reset the display configuration I now boot into OS/2 in VGA. Now however I lose mouse and keyboard almost immediately and don't get them back. 

It looks like both OSes have problems with the emulated hardware. I have no clue where to go from here. Is there any way to debug what it is doing? Are there any tweaks in the settings I can try? Are the problems maybe even know, i.e. the keyboard problem with OS2?

I guess now as Apple silicon is getting more popular more and more people will change from other virtualisation solutions to UTM since it allows to continue running existing x86-based VMs. I have some hope that maybe I am not the only one playting with these exotic OSes?

**Comment by mbert** [comment]
Finally got OS/2 running. Hoping that it will be of interest to others:

I tried through many machine profiles, in the end pc-i440fx-6.2  worked.

Networkadapter: pcnet
Networking: bridged

Display: cirrus-vga
Display driver: SDD (using the "gen" driver rather than the "sdd" driver), so got 1024x768 resolution, not great but good enough

**Reply by oaxjogi** [reply]
Hi,
nice to hear! I had OS/2 and NEXTSTEP runnuning under VirtualBox. Any chance to get OS/2 as a download anywhere? Do you need help for NEXTSTEP/OPENSTEP?

Jogi

**Reply by mbert** [reply]
> Hi, nice to hear! I had OS/2 and NEXTSTEP runnuning under VirtualBox. Any chance to get OS/2 as a download anywhere? 

The older OS/2 could probably be abandonware by now (I still had my original licenses from the 1990s). If I remember correctly, there was a download of OS/2 2.x somewhere, I guess it's just a google search away.

> Do you need help for NEXTSTEP/OPENSTEP?

Actually yes. I have been unable to get my old OPENSTEP installation running. I opened a thread on the nextcomputers.org forum here: http://www.nextcomputers.org/forums/index.php?PHPSESSID=2b1fc5643d774d13c47c5770a6315ebe&topic=4741.0

I still cannot get the system mount the root file system.

**Reply by cbreeves** [reply]
Can you share any details beyond using pc-i440fx-6.2 that you used to get 4.52 up and running?  I'm getting a trap in OS2LVM on booting the install media, so stuck for now.

**Reply by mbert** [reply]
Here's my settings. However I am running WarpServer 3. 
<img width="973" alt="Screenshot 2022-06-24 at 15 55 18" src="https://user-images.githubusercontent.com/383271/175550882-d7795465-b6ff-4868-a236-1f2ed99bc454.png">
<img width="981" alt="Screenshot 2022-06-24 at 15 55 06" src="https://user-images.githubusercontent.com/383271/175550887-39c27439-4e97-4e3e-8e07-df9b0a0670f5.png">
<img width="966" alt="Screenshot 2022-06-24 at 15 54 59" src="https://user-images.githubusercontent.com/383271/175550893-2b625624-b797-4fe5-9ed4-fc0ef36163b7.png">
<img width="971" alt="Screenshot 2022-06-24 at 15 54 50" src="https://user-images.githubusercontent.com/383271/175550895-1c09ee13-a44f-45f3-a11b-9332f789a893.png">

**Reply by hercychang** [reply]
Can you come up with a tutorial, i am also a fan of collecting exotic OS's.

**Reply by tifasoftware** [reply]
I be happy to cook up templates for those in my spare time, if you're interested

**Reply by marc-majka** [reply]
Would love a tutorial for getting OPENSTEP running using UTM on Apple Silicon.  I'm a total VM noob.

**Reply by tifasoftware** [reply]
running OSes like Openstep takes some research like figuring out cpu, machine, graphics, network, and audio.

**Reply by tifasoftware** [reply]
you can also just search up QEMU Openstep and figure it out from there.

**Reply by tifasoftware** [reply]
To install nextstep you have to set the SCSI adapter to the and am53c974, but you can't unless utm implants a SCSI device picker.

**Reply by mbert** [reply]
I was eventually unsuccessful with OPENSTEP. From what I had found out there would have been too many things necessary to do, including even patching qemu. Getting it to run in 86box was rather straightforward. See this thread: https://www.nextcomputers.org/forums/index.php?topic=4741.0

**Reply by melvinlusk** [reply]
Did you ever have any luck with this? I can't get past the OS2LVM error when booting.  I've tried multiple different hardware profiles, as well as different version of OS/2.

**Reply by mbert** [reply]
I got OS/2 WarpServer 4.0 running, see above. I did not try any other version.

**Comment by melvinlusk** [comment]
I had success with OS/2 Warp 4.52 Server.  Here's the settings I used...there may be others that work, but I haven't had a chance to try every single combination.  For an option that's not listed, use the default setting.

Anyone know how to add multiple CPUs to OS/2? I have SMP installed.

Architecture = i386 (x86)
System = Standard PC (i440FX + PIIX, 1996) (pc-i440fx-6.2)
Memory = 512 MB
CPU = Pentium2
Cores = 2 (
OS/2 doesn't accept this, even though I have SMP enabled.
I think it needs two sockets instead of cores...anyone know how to do this in UTM?)	
Force MultiCore = Y
RNG Device = Y
Force PS/2 Controller = Y
USB Support = disabled
Emulated Display Card = Cirrus CLGD 54xx VGA (cirrus-vga)
Add a serial port, otherwise OS/2 will throw erros with COM.SYS when booting.  Choose the defaults.
Network Mode = Bridged (Advanced)
Bridged Interface = Automatic
Emulated Network Card = pcnet
Emulated Audio Card = Creative Sound Blaster 16 (sb16)
Add a floppy drive.
Also add an IDE CD/DVD drive if you're installing from a CD ISO.
Hard drive should use an IDE interface.

**Reply by Samuel525432** [reply]
(For the socket problem) I’m pretty sure that UTM doesn’t have sockets.

**Comment by daanstrik** [comment]
for openstep on ute just disable usb

**Comment by jotagustavo** [comment]
# Installing OS/2 Warp Server on UTM (Apple Silicon M1) — progress report and open issue

Following the setup described in [discussion #3796](https://github.com/utmapp/UTM/discussions/3796), I got much further than expected but I'm stuck on a message-catalog error during install. Posting our findings in case it helps others, and hoping someone recognizes the remaining bug.

## Setup

- Host: Mac M1 (Apple Silicon), UTM + QEMU (TCG software emulation, no HW virtualization for x86 on ARM).
- Media: split boot/system CD pair, not a single combined CD — `os2warpcd1-system.iso` (538.8MB, volume `WARP_4_CP2`) and `os2warpcd2-boot.iso` (552.3MB).
- VM config: x86 (i440fx/PIIX), **Force PS/2 Controller enabled**, USB disabled, serial port enabled, network = pcnet, HD = qcow2 5GB, one CD-ROM drive created as **Removable** from the start.

## Bugs found and fixed along the way

- **Keyboard completely dead at boot.** Caused by `i8042=off` on the machine and/or `-S` (CPU started paused) when driving QEMU manually. Fixed via UTM's built-in **Force PS/2 Controller** toggle — don't fight this one manually, use the GUI option.
- **"OS/2 is unable to operate your hard disk or diskette drive."** Looked like a disk geometry/CHS issue, wasn't — it went away once the keyboard/PS2 fix above was applied.
- **CD-ROM hangs forever on "Loading disks..."** Confirmed via QEMU monitor (`info registers`) as a CPU HLT waiting on an IDE/ATAPI IRQ that never comes. Root cause was the old single-CD media (El Torito boot catalog incompatibility). Switched to the split boot/system CD pair above and this went away entirely — no floppy/vvfat workarounds needed.
- **`Bus 'ide.2' not found`** when trying to attach a second CD-ROM. UTM's simple "New IDE Drive" GUI doesn't support master/slave pairing, so you're capped at 2 IDE devices total via the GUI. Workaround: keep HD + 1 CD-ROM only, and swap ISOs live via the **Virtual Machine → Drives** menu instead of adding a drive.
- **Live CD swap not detected** (`dir d:` kept showing the old disc's contents after using "Change"). Fix: always **Eject** first, confirm the drive errors out, then **Change** to the new ISO.
- **"Removable" checkbox greyed out.** This property can only be set when the drive is created. If you forgot, delete the drive and recreate it with Removable checked before importing the ISO.
- **"Volumes Too Small"** (OS/2 wants ≥120MB, 450MB for full install). Fixed by resizing the HD to 5GB in UTM and recreating the C: volume via the Logical Volume Management Tool.

## Where we're stuck

During install, running:
```
d:
cd \cid\exe\os2
unpack2 d:\os2image\disk_2\BUNDLE c:\test
```
fails partway through (around extracting `NAMEMOVE.EXE`) with:
```
SYS0318: Message file OSO001.MSG cannot be found for message 3175
```

`OSO001.MSG` genuinely isn't present on `disk_2` — it only exists on `disk_3` and `disk_6`. Two things we tried:

1. Rebuilt the system ISO with `OSO001.MSG` copied into the `disk_2` folder (via `hdiutil makehybrid`). No effect.
2. Manually copied `OSO001.MSG` to `C:\` and `C:\OS2\` from inside the running VM. This worked **once** — install progressed further and hit a different crash (`SYS1811`, Trap 000E / page fault, not yet investigated). On a later attempt, with the exact same file confirmed present at `C:\OSO001.MSG` (175354 bytes, verified via `dir`), the identical `SYS0318` / message 3175 error came back.

So the fix is not reliably reproducible even with the file demonstrably in place, which means the real search/lookup path for OS/2's message catalog isn't what we assumed. Next thing we're going to try is consolidating everything `UNPACK2.EXE` might need (disk_2, disk_3, disk_6 contents, including `OSO001.MSG`) into one single directory before running the extraction, instead of relying on the CD in the drive.

**If anyone has run into `SYS0318`/message 3175 during a Warp Server MCP1 install, or knows how OS/2's message-catalog lookup actually resolves paths (it's not simply same-folder-as-caller or `DPATH`), we'd appreciate any pointers.**
